HW2/P4/driver.py

```diff
@@ -19,6 +19,46 @@ def py_median_3x3(image, iterations=10, num_threads=1):
     ''' repeatedly filter with a 3x3 median '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
+    num_lines = tmpA.shape[0]
+    #list of lists where lines_by_tid[tid] = list of lines that tid is responsible
+    # e.g. tid == 1 ==> lines_by_tid[1] = [1 + num_threads * 0, 1 + num_threads * 1, 1 + num_threads * 2,...]
+    '''
+    lines_by_tid = []
+    for tid in range(num_threads):
+        #tid is thread_id
+        tid_lines = []
+        # this will fail if num_threads > # of lines
+        for x in range(num_lines/num_threads):
+            tid_lines.append(tid + num_threads * x)
+        lines_by_tid.append(tid_lines)
+    #if # lines is not divisible by num threads, distribute remaining lines
+    x = num_lines/num_threads
+    for tid in range(num_lines%num_threads):
+        lines_by_tid[tid].append(tid + num_threads * x)
+    '''  
+    #print threading.active_count()
+    for i in range(iterations):
+        threads = []
+        for tid in range(num_threads):
+            #offset = tid, step = num_threads
+            t = threading.Thread(target=filtering.median_3x3,args=(tmpA, tmpB, tid, num_threads))
```

Comment:
I might be missing something here, but what exactly is your locking mechanism here? How do you make sure the invariant of threads n-1 and n+1 only starting when n is complete is upheld? Does this code actually pass the assertion test in line 97 below?